### ISSUE TITLE
Use different client key shares for ESNI and TLS KE?

### ISSUE THREAD

**Issue by ekr** [ekr/draft-rescorla-tls-esni#67]
Using the same key share looked like a good idea, but it's actually a real pain to implement because you need to get pretty involved with the TLS 1.3 key exchange.

Another approach would be to have separate keys but bind the ClientHello.KeyShare extension into the ESNI extension. I had thought of putting a hash in the extension plaintext but @martinthomson points out that if you use it as AAD, then the client-facing server will automatically detect mismatch

**Comment by kazuho** [comment]
> Using the same key share looked like a good idea, but it's actually a real pain to implement because you need to get pretty involved with the TLS 1.3 key exchange.

I am not sure if are going to have such pain in general. In my view, what you can do is as follows:
* construct encrypted_server_name extension
* construct ClientHello, with passing in the ESNI extension and the key-exchange context that was used for constructing the extension as the arguments
  * add the key-exchange context being passed in to ClientHello
  * omit constructing a key-exchange context if the id of the context to-be-constructed matches that of the supplied context

To me, they seem to be fairly limited change.

Generally speaking, I think the question of if we should have different key shares depends on if we want to apply different crypto to the ESNI and the TLS application payload. My preference is to use single crypto whenever possible for simplicity and in hope that we can reuse ESNIKeys for other purposes in the future.

**Comment by kazuho** [comment]
> I had thought of putting a hash in the extension plaintext but @martinthomson points out that if you use it as AAD, then the client-facing server will automatically detect mismatch

:+1:

**Comment by ekr** [comment]
Sorry, there are two values here:

1. The ESNIKeys hash, which I agree has to be in the clear.
2. The hash of KeyShares to prevent substitution attacks.

@kazuho: I agree it's doable, as I just got my POC working. Yes, the CH is as you say, which is modestly disruptive. On the server side it's a bit disruptive because of the way NSS handles extensions, which is to say in-order. So, ESNI could come either before or after KeyShares. I could refactor this of course.

It also affects negotiation, so if the client didn't pick the server's most preferred group, you have to refrain from sending HRR, which is a bit of a pain.

**Comment by kazuho** [comment]
Thank you for the clarification.

I can understand the pain on the server-side. Current approach of ESNI changes the negotiation from client-offers-a-list-server-chooses-one, to server-offers-a-list-client-chooses-one.

I think that it might be fine to revert to the TLS 1.3 way of negotiation. Actually we might be required to do that, if we want to prevent a client-facing server downgrading the crypto in the split server scenario.

OTOH, I am not sure if we would want to use a _separate_ key share for ESNI, because that would require additional CPU cost. My benchmark tells me that in case of X25519, key generation takes about half the time of key derivation. That means that if we are going to have separate key shares but the algorithm collides, the client will be spending 17% more CPU cycles (i.e. 1 key generation + 2 derivations vs. 2 key generations and 2 derivations).

Considering the cost, I think that we'd be tempted to try reuse the key-share for two purposes (i.e. ESNI and TLS handshake), even if we define separate key exchange fields for the two.

And that in turn make me wonder what the value is for having separate key exchange fields.

To summarize, I do not oppose to allowing client send multiple key shares or allowing server send HRR when ESNI is involved, but I prefer using the value of the KeyShares extension for carrying the ECDH key being used to encrypt ESNI.

**Comment by ekr** [comment]
Fixed
